pkg/controller/storagenode/storagenode.go

```diff
@@ -0,0 +1,393 @@
+		return err
+	}
+
+	if err := c.syncStorage(cluster, storageNode); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Controller) syncKVDB(
+	cluster *corev1alpha1.StorageCluster,
+	storageNode *corev1alpha1.StorageNode,
+) error {
+	if cluster.Spec.Kvdb != nil && !cluster.Spec.Kvdb.Internal {
+		return nil
+	}
+
+	// ensure kvdb pods are present on nodes running kvdb
+	// list kvdb nodes on the node for this cluster
+	kvdbPodList := &v1.PodList{}
+	fieldSelector := fields.SelectorFromSet(map[string]string{"nodeName": storageNode.Name})
+	err := c.client.List(context.TODO(), kvdbPodList, &client.ListOptions{
+		Namespace:     storageNode.Namespace,
+		LabelSelector: labels.SelectorFromSet(c.kvdbPodLabels(cluster)),
+		FieldSelector: fieldSelector,
+	})
+	if err != nil {
+		return err
+	}
+
+	if isNodeRunningKVDB(storageNode) { // create kvdb pod if not present
+		node := &v1.Node{}
+		isBeingDeleted := false
+		err = c.client.Get(context.TODO(), client.ObjectKey{Name: storageNode.Name}, node)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				c.log(storageNode).Debugf("Kubernetes node is no longer present")
+				isBeingDeleted = true
+			} else {
+				c.log(storageNode).Warnf("failed to get node: %s due to: %s", storageNode.Name, err)
+			}
+		} else {
+			isBeingDeleted, err = k8s.IsNodeBeingDeleted(node, c.client)
+			if err != nil {
+				c.log(storageNode).Warnf("failed to check if node: %s is being deleted due to: %v", node.Name, err)
+			}
+		}
+
+		if !isBeingDeleted && len(kvdbPodList.Items) == 0 {
+			pod, err := c.createKVDBPod(cluster, storageNode)
+			if err != nil {
+				return err
+			}
+
+			c.log(storageNode).Infof("creating kvdb pod: %s/%s", pod.Namespace, pod.Name)
+			err = c.client.Create(context.TODO(), pod)
+			if err != nil {
+				return err
+			}
+		}
+	} else { // delete pods if present
+		for _, p := range kvdbPodList.Items {
+			c.log(storageNode).Debugf("deleting kvdb pod: %s/%s", p.Namespace, p.Name)
+			err = c.client.Delete(context.TODO(), &p)
+			if err != nil {
+				c.log(storageNode).Warnf("failed to delete pod: %s/%s due to: %v", p.Namespace, p.Name, err)
+			}
+		}
+	}
+	return nil
+}
+func (c *Controller) syncStorage(
+	cluster *corev1alpha1.StorageCluster,
+	storageNode *corev1alpha1.StorageNode,
+) error {
+	// sync the storage labels on pods
+	portworxPodList := &v1.PodList{}
+	pxLabels := c.Driver.GetSelectorLabels()
+	err := c.client.List(
+		context.TODO(),
+		portworxPodList,
+		&client.ListOptions{
+			Namespace:     storageNode.Namespace,
+			LabelSelector: labels.SelectorFromSet(pxLabels),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get list of portworx pods. %v", err)
+	}
+
+	for _, pod := range portworxPodList.Items {
+		podCopy := pod.DeepCopy()
+		controllerRef := metav1.GetControllerOf(podCopy)
+		if controllerRef != nil && controllerRef.UID == cluster.UID &&
+			len(pod.Spec.NodeName) != 0 && storageNode.Name == podCopy.Spec.NodeName {
```

Comment:
We should ignore pod which is marked for deletion - `pod.DeletionTimestamp != nil`
Although storage cluster controller only starts a pod if the previous is completely terminated, just to be safe we can add that check, so update the newer pod and not the one in terminating state.